subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformInfo.java

```diff
@@ -162,46 +147,130 @@ public void resolveDependencies(TaskDependencyResolver dependencyResolver, Actio
         private Set<WorkInfo> getDependencies(TaskDependencyResolver dependencyResolver) {
             return dependencyResolver.resolveDependenciesFor(null, artifactSet.getBuildDependencies());
         }
+
+        private class InitialArtifactTransformationStepOperation implements RunnableBuildOperation {
+            private List<File> result;
+            private Throwable failure;
+            private final BuildOperationExecutor buildOperationExecutor;
+
+            public InitialArtifactTransformationStepOperation(BuildOperationExecutor buildOperationExecutor) {
+                this.buildOperationExecutor = buildOperationExecutor;
+            }
+
+            @Override
+            public BuildOperationDescriptor.Builder description() {
+                String displayName = "Transform " + artifactSet.getArtifactId().getDisplayName() + " with " + artifactTransformer.getDisplayName();
+                return BuildOperationDescriptor.displayName(displayName)
+                    .progressDisplayName(displayName)
+                    .operationType(BuildOperationCategory.TRANSFORM);
+            }
+
+            @Override
+            public void run(BuildOperationContext context) {
+                ResolveArtifacts resolveArtifacts = new ResolveArtifacts(artifactSet);
+                buildOperationExecutor.runAll(resolveArtifacts);
+                ResolvedArtifactCollectingVisitor visitor = new ResolvedArtifactCollectingVisitor();
+                resolveArtifacts.getResult().visit(visitor);
+                Set<Throwable> failures = visitor.getFailures();
+                if (!failures.isEmpty()) {
+                    Throwable failure;
+                    if (failures.size() == 1 && Iterables.getOnlyElement(failures) instanceof ResolveException) {
+                        failure = Iterables.getOnlyElement(failures);
+                    } else {
+                        failure = new DefaultLenientConfiguration.ArtifactResolveException("artifacts", artifactTransformer.getDisplayName(), "artifact transform", failures);
+                    }
+                    this.failure = failure;
+                    this.result = Collections.emptyList();
+                    return;
+                }
+                ResolvedArtifactResult artifact = Iterables.getOnlyElement(visitor.getArtifacts());
+
+                TransformArtifactOperation operation = new TransformArtifactOperation(artifact.getId(), artifact.getFile(), artifactTransformer, BuildOperationCategory.UNCATEGORIZED);
```

Comment:
This class seems unnecessary now. E.g. it has an unused `description` method for which you need to pass a category.

Reply:
The class is still used when running artifact transforms "ad-hoc", i.e. as part of task execution. I could also run the operation as a sub-operation of the top-level transform, so we would see some progress for each individual file. The `category` parameter to the build operation can be removed.

I decided not to since the current logging seems to be enough.